sql/database.go

```diff
@@ -292,16 +330,194 @@ func Open(uri string, opts ...Opt) (*sqliteDatabase, error) {
 	return db, nil
 }
 
+func ensureDBSchemaUpToDate(logger *zap.Logger, db *sqliteDatabase, config *conf) (*sqliteDatabase, error) {
+	before, after, err := config.schema.CheckDBVersion(logger, db)
+	switch {
+	case err != nil:
+		return db, err
+	case before == after:
+		return db, nil
+	case before > after:
+		// TODO: this should be logged by the caller
+		logger.Error("database version is newer than expected - downgrade is not supported",
+			zap.Int("current version", before),
+			zap.Int("target version", after),
+		)
+		return db, fmt.Errorf("%w: %d > %d", ErrTooNew, before, after)
+	case !config.enableMigrations:
+		// TODO: this should be logged by the caller
+		logger.Error("database version is too old",
+			zap.Int("current version", before),
+			zap.Int("target version", after),
+		)
+		return db, fmt.Errorf("%w: %d < %d", ErrOldSchema, before, after)
+	case config.temp:
+		// Temporary database, do migrations without transactions
+		// and sync afterwards
+		return db, config.schema.MigrateTempDB(logger, db, before)
+	case config.vacuumState != 0 &&
+		before <= config.vacuumState &&
+		strings.HasPrefix(config.uri, "file:"):
+		logger.Info("running migrations",
+			zap.Int("current version", before),
+			zap.Int("target version", after),
+		)
+		return db.copyMigrateDB(config)
+	default:
+		// Do not produce extra "running migrations" log message for the
+		// temporary DB, as it was already logged
+		if !config.temp {
+			logger.Info("running migrations in-place",
+				zap.Int("current version", before),
+				zap.Int("target version", after),
+			)
+		} else {
+			logger.Info("applying migrations to temporary DB",
+				zap.Int("current version", before),
+				zap.Int("target version", after),
+			)
+		}
+		return db, config.schema.Migrate(logger, db, before, config.vacuumState)
+	}
+}
+
 func Version(uri string) (int, error) {
 	pool, err := sqlitex.Open(uri, sqlite.SQLITE_OPEN_READONLY, 1)
 	if err != nil {
 		return 0, fmt.Errorf("open db %s: %w", uri, err)
 	}
 	db := &sqliteDatabase{pool: pool}
-	defer db.Close()
-	return version(db)
+	v, err := version(db)
+	return v, errors.Join(err, db.Close())
```

Comment:
not sure why `errors.Join` is needed. This is reinventing the idiomatic go error checking with `if` blocks and error instrumentation. IMO it would be good to stick to that pattern.

Reply:
Removed `errors.Join`